File: TestLatex.java
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class TestLatex {
    static Map<Character, String> superMap = new HashMap<>();
    static Map<Character, String> subMap = new HashMap<>();

    static {
        superMap.put('0', "\u2070"); superMap.put('1', "\u00B9"); superMap.put('2', "\u00B2");
        superMap.put('+', "\u207A"); superMap.put('-', "\u207B"); superMap.put('=', "\u207C");
        superMap.put('x', "\u02E3"); superMap.put('n', "\u207F");
        
        subMap.put('0', "\u2080"); subMap.put('1', "\u2081"); subMap.put('2', "\u2082");
        subMap.put('+', "\u208A"); subMap.put('-', "\u208B"); subMap.put('=', "\u208C");
        subMap.put('x', "\u2093"); subMap.put('n', "n");
    }

    static String toSup(String s) {
        StringBuilder sb = new StringBuilder();
        for (char c : s.toCharArray()) sb.append(superMap.getOrDefault(c, String.valueOf(c)));
        return sb.toString();
    }

    static String toSub(String s) {
        StringBuilder sb = new StringBuilder();
        for (char c : s.toCharArray()) sb.append(subMap.getOrDefault(c, String.valueOf(c)));
        return sb.toString();
    }

    static String latexSymbolToStr(String sym) {
        return sym.replace("\\infty", "\u221E");
    }

    public static void main(String[] args) {
        String s = "\\int_{-\\infty}^{\\infty} e^{-x^2} dx = \\sqrt(\\pi)\n\\sum_{n=0}^{\\infty} r^n = \\frac{1}{1-r}";
        
        s = s.replace("\\left", "").replace("\\right", "");
        s = s.replace("\\{", "{").replace("\\}", "}");
        
        Matcher m1 = Pattern.compile("\\^\\{([^}]*)\\}").matcher(s);
        StringBuffer sb1 = new StringBuffer();
        while (m1.find()) {
            m1.appendReplacement(sb1, Matcher.quoteReplacement(toSup(latexSymbolToStr(m1.group(1)))));
        }
        m1.appendTail(sb1);
        s = sb1.toString();
        
        Matcher m2 = Pattern.compile("_\\{([^}]*)\\}").matcher(s);
        StringBuffer sb2 = new StringBuffer();
        while (m2.find()) {
            m2.appendReplacement(sb2, Matcher.quoteReplacement(toSub(latexSymbolToStr(m2.group(1)))));
        }
        m2.appendTail(sb2);
        s = sb2.toString();

        s = s.replace("\\int", "\u222B").replace("\\sum", "\u2211").replace("\\infty", "\u221E");
        
        System.out.println(s);
    }
}

File: test_latex.kt
import java.io.File

fun main() {
    val superMap = mapOf(
        '0' to "\u2070", '1' to "\u00B9", '2' to "\u00B2", '3' to "\u00B3", '4' to "\u2074",
        '5' to "\u2075", '6' to "\u2076", '7' to "\u2077", '8' to "\u2078", '9' to "\u2079",
        '+' to "\u207A", '-' to "\u207B", '=' to "\u207C", '(' to "\u207D", ')' to "\u207E",
        'n' to "\u207F", 'i' to "\u2071",
        'x' to "\u02E3", 'y' to "\u02B8", 'a' to "\u1D43", 'b' to "\u1D47", 'c' to "\u1D9C"
    )

    val subMap = mapOf(
        '0' to "\u2080", '1' to "\u2081", '2' to "\u2082", '3' to "\u2083", '4' to "\u2084",
        '5' to "\u2085", '6' to "\u2086", '7' to "\u2087", '8' to "\u2088", '9' to "\u2089",
        '+' to "\u208A", '-' to "\u208B", '=' to "\u208C", '(' to "\u208D", ')' to "\u208E",
        'a' to "\u2090", 'e' to "\u2091", 'o' to "\u2092", 'i' to "\u1D62", 'u' to "\u1D64",
        'x' to "\u2093", 'r' to "\u1D63", 'v' to "\u1D65", 'j' to "\u2C7C"
    )

    fun toSup(s: String) = s.map { superMap[it] ?: it.toString() }.joinToString("")
    fun toSub(s: String) = s.map { subMap[it] ?: it.toString() }.joinToString("")

    fun latexSymbolToStr(sym: String): String {
        var r = sym
        r = r.replace("\\infty", "\u221E")
        return r
    }

    var s = "\\int_{-\\infty}^{\\infty} e^{-x^2} dx = \\sqrt(\\pi)\n\\sum_{n=0}^{\\infty} r^n = \\frac{1}{1-r}, \\quad |r| < 1"
    
    // Step 1
    s = s.replace("\\left", "").replace("\\right", "")
    s = s.replace("\\{", "{").replace("\\}", "}")
    s = s.replace("\\,", " ").replace("\\!", "").replace("\\;", " ").replace("\\:", " ")
    s = s.replace("\\\\", "\n")

    // Step 4
    s = s.replace(Regex("\\^\\{([^}]*)\\}")) { m ->
        val inner = latexSymbolToStr(m.groupValues[1])
        toSup(inner)
    }
    s = s.replace(Regex("_\\{([^}]*)\\}")) { m ->
        val inner = latexSymbolToStr(m.groupValues[1])
        toSub(inner)
    }
    s = s.replace(Regex("\\^([0-9a-zA-Z+\\-])")) { m -> toSup(m.groupValues[1]) }
    s = s.replace(Regex("_([0-9a-zA-Z+\\-])")) { m -> toSub(m.groupValues[1]) }

    // Step 5
    s = s.replace("\\int", "\u222B")
    s = s.replace("\\sum", "\u2211")
    s = s.replace("\\infty", "\u221E")

    println(s)
}
